int16 distance
int8 p
int8 d
int8 i
